src/qutip_qip/circuit/circuit.py

```diff
@@ -588,9 +589,6 @@ def resolve_gates(self, basis=["CNOT", "RX", "RY", "RZ"]):
                     basis_1q.append(gate)
                 else:
                     pass
-                    raise NotImplementedError(
```

Comment:
Unrelated to this PR, but there is a lot of strange magic here in the construction of the list of basis operations. I don't know why. Do you perhaps?

Reply:
Yeah, you are right that there is some complicated logic here.

The goal is to decompose those gates to the native basis gates (like SWAP to CNOT), unless it is already provided in the basis as a native gates, such as `RZX` (e.g. for Hamiltonian simulation).

`basis_2q_valid` and `basis_1q_valid` contain the basis gate list of which decomposition is implemented in qutip. Previously, if `RZX` is provided as a native gate the algorithm raises a not-implemented error, because it does not know how to decompose other gates to `RZX`. But now we want it to just pass the RZX gate on without trying to decompose any other gates (like CNOT) to it.

The overall logic of the decomposition can be drastically simplified if it is refactored, since now we have the gate class. The decomposition scheme can be defined under each gate class, not independently. But this will take some work.

Reply:
Thanks for the great explanation. Would be cool to sort this out, but definitely not for here. :)